### PR TITLE
Flush ObjC cached channels when network connectivity changes

### PR DESCRIPTION
**Pull request by muxi**
Follow-up of Issue #4064 

cc: @makdharma

### CODE REVIEW

File: src/objective-c/GRPCClient/GRPCCall.m [hunk 1]
```diff
@@ -397,15 +383,35 @@ - (void)startWithWriteable:(id<GRXWriteable>)writeable {
...
     if (strongSelf) {
       [strongSelf finishWithError:[NSError errorWithDomain:kGRPCErrorDomain
                                                       code:GRPCErrorCodeUnavailable
                                                   userInfo:@{NSLocalizedDescriptionKey: @"Connectivity lost."}]];
-      [[GRPCHost hostWithAddress:strongSelf->_host] disconnect];
     }
-  }];
+  };
+  [_connectivityMonitor handleLossWithHandler:handler
+                      wifiStatusChangeHandler:handler];
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/GRPCCall.m]: But we don't want to cancel pending RPCs when WiFi comes and goes, no?

File: src/objective-c/GRPCClient/private/GRPCConnectivityMonitor.m [hunk 2]
```diff
@@ -129,6 +130,13 @@ - (nullable instancetype)initWithReachability:(nullable SCNetworkReachabilityRef
   if ((self = [super init])) {
     _reachabilityRef = CFRetain(reachability);
     _queue = dispatch_get_main_queue();
+    SCNetworkReachabilityFlags flags;
+    if (SCNetworkReachabilityGetFlags(_reachabilityRef, &flags)) {
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCConnectivityMonitor.m]: I read in some Apple documentation that `SCNetworkReachabilityGetFlags` can block, if it needs to resolve DNS. So we'd need to either ensure that this isn't called in the main thread (I suspect we can't), or just initialize to `nil`.

File: src/objective-c/GRPCClient/private/GRPCConnectivityMonitor.m [hunk 3]
```diff
@@ -129,6 +130,13 @@ - (nullable instancetype)initWithReachability:(nullable SCNetworkReachabilityRef
...
+      _previousReachabilityFlags =
+          [[GRPCReachabilityFlags alloc] initWithFlags:flags];
+    } else {
+      _previousReachabilityFlags = 0;
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCConnectivityMonitor.m]: `nil`, it's a pointer :) if the compiler didn't complain, the compiler is bad.

File: src/objective-c/GRPCClient/private/GRPCHost.m [hunk 4]
```diff
@@ -98,6 +100,16 @@ - (nullable instancetype)initWithAddress:(NSString *)address {
       _address = address;
       _secure = YES;
       kHostCache[address] = self;
+
+      if (!connectivityMonitor) {
+        connectivityMonitor =
+            [GRPCConnectivityMonitor monitorWithHost:hostURL.host];
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCHost.m]: This is initialized with the specific host name, but you're setting it to a static variable?

File: src/objective-c/GRPCClient/private/GRPCReachabilityFlagNames.xmacro.h [hunk 5]
```diff
@@ -65,3 +65,4 @@ GRPC_XMACRO_ITEM(interventionRequired, InterventionRequired)
 GRPC_XMACRO_ITEM(connectionOnDemand, ConnectionOnDemand)
 GRPC_XMACRO_ITEM(isLocalAddress, IsLocalAddress)
 GRPC_XMACRO_ITEM(isDirect, IsDirect)
+GRPC_XMACRO_ITEM(isWWAN, IsWWAN)
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCReachabilityFlagNames.xmacro.h]: I was surprised that a case was missing, but it's hidden above within the `#if TARGET_OS_IPHONE`. It looks like giving it a different name wasn't a good idea :)

File: src/objective-c/tests/Connectivity/Podfile [hunk 6]
```diff
@@ -1,10 +1,31 @@
 install! 'cocoapods', :deterministic_uuids => false
+platform :ios, '8.0'
 
 # Location of gRPC's repo root relative to this file.
 GRPC_LOCAL_SRC = '../../../..'
 
 target 'ConnectivityTestingApp' do
   pod 'gRPC', :path => GRPC_LOCAL_SRC
+  pod 'gRPC-Core', :path => GRPC_LOCAL_SRC
+  pod 'gRPC-ProtoRPC', :path => GRPC_LOCAL_SRC
```
Comment (jcanizales) [inline src/objective-c/tests/Connectivity/Podfile]: You want `'gRPC-RxLibrary'` too.

File: src/objective-c/GRPCClient/GRPCCall.m (hunk 1 above)
Comment (muxi) [inline src/objective-c/GRPCClient/GRPCCall.m]: I tried that; got some problem in that way. Apple seems to force you to switch to wifi even if cellular is still available.

File: src/objective-c/GRPCClient/private/GRPCConnectivityMonitor.m (hunk 2 above)
Comment (muxi) [inline src/objective-c/GRPCClient/private/GRPCConnectivityMonitor.m]: In that case it would be a problem. I'll just initialize it to nil then.

File: src/objective-c/GRPCClient/private/GRPCHost.m (hunk 4 above)
Comment (muxi) [inline src/objective-c/GRPCClient/private/GRPCHost.m]: Yea. I think the host name does not matters here and assuming all the host names have the same reachability (we talked about this yesterday I remember). I don't need ReachabilityRef to all the hosts. As long as there is at least one host cached, I'm gonna need one ReachabilityRef to flush the channel.

File: src/objective-c/GRPCClient/GRPCCall.m (hunk 1 above)
Comment (jcanizales) [inline src/objective-c/GRPCClient/GRPCCall.m]: What was the problem?

File: src/objective-c/GRPCClient/private/GRPCReachabilityFlagNames.xmacro.h (hunk 5 above)
Comment (muxi) [inline src/objective-c/GRPCClient/private/GRPCReachabilityFlagNames.xmacro.h]: missed :(

File: src/objective-c/GRPCClient/private/GRPCHost.m (hunk 4 above)
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCHost.m]: Ok, but then you need to (1) document that, so people don't go "huh??", and (2) think about concurrent writes and reads to that static variable.

File: src/objective-c/GRPCClient/GRPCCall.m (hunk 1 above)
Comment (muxi) [inline src/objective-c/GRPCClient/GRPCCall.m]: This call keeps waiting and next call does not get started.

Comment (jcanizales) [inline src/objective-c/GRPCClient/GRPCCall.m]: Wow, that's super wrong behavior. If after that you turn WiFi off again, does the initial call end? Does the subsequent call start?

File: src/objective-c/GRPCClient/private/GRPCHost.m (hunk 4 above)
Comment (muxi) [inline src/objective-c/GRPCClient/private/GRPCHost.m]: Would that @synchronized(kHostCache) be good enough for concurrency? in both initWithAddress and flushChannelCache

File: src/objective-c/GRPCClient/private/GRPCReachabilityFlagNames.xmacro.h (hunk 5 above)
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCReachabilityFlagNames.xmacro.h]: I'd suggest renaming that one to `isWWAN`, so next time someone goes looking for it it's easier to spot.

File: src/objective-c/GRPCClient/GRPCCall.m (hunk 1 above)
Comment (muxi) [inline src/objective-c/GRPCClient/GRPCCall.m]: Actually I think the problem was due to another issue. After fixing now it seems working.

**Comment by muxi** [comment]
test this please

**Comment by muxi** [comment]
@jcanizales request another look of review :)

File: src/objective-c/GRPCClient/private/GRPCConnectivityMonitor.m [hunk 7]
```diff
@@ -58,45 +58,50 @@ - (BOOL)isCell { \
...
 
-  if (self.isCell) {
-    [activeOptions addObject:@"isCell"];
-  }
+if (self.isCell) {
+  [activeOptions addObject:@"isCell"];
+}
 
-   */
+ */
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCConnectivityMonitor.m]: If clang-format is going to align this comment with the following `#define`, then both things should be indented, being part of the `description` method. Do you want to separate these things in a different PR? They seem to require plenty of review, and are irrelevant to the feature which is kind of urgent IIUC.

File: src/objective-c/GRPCClient/GRPCCall.m [hunk 8]
```diff
@@ -375,37 +375,43 @@ - (void)startWithWriteable:(id<GRXWriteable>)writeable {
...
     typeof(self) strongSelf = weakSelf;
     if (strongSelf) {
       [strongSelf finishWithError:[NSError errorWithDomain:kGRPCErrorDomain
                                                       code:GRPCErrorCodeUnavailable
                                                   userInfo:@{NSLocalizedDescriptionKey: @"Connectivity lost."}]];
-      [[GRPCHost hostWithAddress:strongSelf->_host] disconnect];
     }
-  }];
+  };
+  [_connectivityMonitor handleLossWithHandler:handler
```
Comment (muxi) [inline src/objective-c/GRPCClient/GRPCCall.m]: yep that's a problem. I'm gonna need to set the monitor earlier than creating the channel

Comment (muxi) [inline src/objective-c/GRPCClient/GRPCCall.m]: It seems to be ok. The host would flush right away and later when channel is created a new channel is created. Is it?

File: src/objective-c/GRPCClient/private/GRPCHost.m (hunk 4 above)
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCHost.m]: It is! So just add a comment on top of the static variable declaration, like
```objective-c
// Thread safety guarded by @synchronized(kHostCache).
```

File: src/objective-c/GRPCClient/private/GRPCHost.m [hunk 9]
```diff
@@ -99,6 +105,16 @@ - (nullable instancetype)initWithAddress:(NSString *)address {
       _secure = YES;
       kHostCache[address] = self;
     }
+    // Keep a single monitor to flush the cache if the connectivity status changed
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCHost.m]: nit: s/changed/changes.

Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCHost.m]: Also explain why we're doing this, for the people of the future!

File: src/objective-c/GRPCClient/private/GRPCHost.m [hunk 10]
```diff
@@ -54,6 +54,10 @@
 #define GRPC_OBJC_VERSION_STRING @"1.0.0"
 
 static NSMutableDictionary *kHostCache;
+
+// This connectivity monitor flushes the host cache when connectivity status
+// changes. Otherwise, a new call might still use the cached channel which is
+// no longer available and will call gRPC to hang
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCHost.m]: Well, the channel is available, and WiFi coming up doesn't mean we've lost cell connectivity!

File: src/objective-c/GRPCClient/private/GRPCHost.m [hunk 11]
```diff
@@ -106,6 +111,7 @@ - (nullable instancetype)initWithAddress:(NSString *)address {
       kHostCache[address] = self;
     }
     // Keep a single monitor to flush the cache if the connectivity status changed
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCHost.m]: nit: s/changed/changes.
